pkg/planner/util/null_misc.go

```diff
@@ -68,62 +157,233 @@ func isNullRejectedInList(ctx base.PlanContext, expr *expression.ScalarFunction,
 			return true
-		} else if isTrue, err := x.Value.ToBool(sc.TypeCtxOrDefault()); err == nil && isTrue == 0 {
+		}
+		for _, a := range x.GetArgs() {
+			if containsOpaqueInfoFunc(a) {
+				return true
+			}
+		}
+		return false
+	default:
+		return false
+	}
+}
+
+// tryNullifyInnerAndFold substitutes every column that belongs to `inner` with NULL,
+// then folds the expression, and applies 3VL on the folded result.
+// This is the semantic core of "null-rejecting" detection for outer-to-inner join elimination:
+// if P(outer, NULL) can never be TRUE (i.e., folds to NULL or FALSE), then P is null-rejecting.
+func tryNullifyInnerAndFold(
+	ctx base.PlanContext,
+	inner *expression.Schema,
+	pred expression.Expression,
+	skipPlanCacheCheck bool,
+) (decided bool, reject bool) {
+	// Guard: bail out if the predicate mentions any non-inner column.
+	hasNonInner := false
+	for _, c := range expression.ExtractColumns(pred) {
+		if !inner.Contains(c) {
+			hasNonInner = true
+			break
+		}
+	}
+	// Only enter the expensive path when "only referencing inner columns"; if info functions appear at this point,
+	// directly abandon giving conclusions through folding (more conservative).
+	if !hasNonInner {
+		if containsOpaqueInfoFunc(pred) {
+			return false, false
+		}
+	}
+	buildCtx := ctx.GetNullRejectCheckExprCtx()
+	folded, err := expression.EvaluateExprWithNull(buildCtx, inner, pred, skipPlanCacheCheck)
+	if err != nil || folded == nil {
+		return false, false
+	}
+
+	if hasNonInner {
+		// Do not trust a constant produced by folding when outer columns exist.
+		// Example: (outer AND inner!=NULL) IS FALSE  → still depends on outer.
+		return false, false
+	}
+
+	// Safe: the predicate was over inner-only columns, so P(NULL) is a real constant.
+	return judgeFoldedConstant(ctx, folded)
+}
+
+// +------------------------+
+// | Constant?             |--Yes--> NULL/FALSE? → true
+// +------------------------+                     TRUE → false
+//	          |
+//	          No
+//	          v
+// +------------------------+
+// | All-constant tree?     |--Yes--> Fold → NULL/FALSE? → true
+// +------------------------+                     TRUE → false
+//	          |
+//	          No
+//	          v
+// +------------------------+
+// | Structural decision?   |--Yes--> return decision
+// +------------------------+
+//	          |
+//	          No
+//	          v
+// +------------------------+
+// | Ref inner columns?     |--No--> false
+// +------------------------+
+//	          |
+//	          Yes
+//	          v
+// +------------------------+
+// | Replace inner→NULL &   |
+// | constant fold          |--Fold→ NULL/FALSE? → true
+// +------------------------+         else → false
+
+// isNullRejectedSimpleExpr decides whether `predicate` is null-rejecting w.r.t. the `innerSchema`.
+// Strategy (fast → slow):
+//  1. Literal constant fast-path: WHERE NULL / 0 / 1.
+//  2. Whole constant tree: FoldConstant, then judge folded constant.
```

Comment:
For case 1 and 2, why can we say they are null-rejective?
For example, in the query `select * from t1 left join t2 on t1.a=t2.a where null and (1+2)=3`, can we say `null` and `(1+2)=3` are null-rejective for this query?

Reply:
For NULL AND (1+2)=3:
- (1+2)=3 is TRUE.
- NULL AND TRUE is NULL under three-valued logic (3VL).
In a WHERE clause, NULL is treated as FALSE, so it rejects all rows.